### PR TITLE
1141 revoke share fix

### PR DESCRIPTION
**Pull request by SofiaSazonova**
### Feature or Bugfix
- Bugfix

### Detail
- Health check is now started with check if the principal role exists
- If Role is missing:
-- add share item will fail (Share failed) and then the user will be able delete it
-- revoke of the item will be successful ( SHARE_SUCCEEDED) 
- If Role was missing, but then restored with the same name and attached policies (to keep arn), the share can be revoked and applied again successfuly

### Relates
- [1141](https://github.com/data-dot-all/dataall/issues/1141)

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by SofiaSazonova** [comment]
How to test:

- [ ] Create consumption role for the env and share request for this role
- [ ] After all items are shared successfully, delete IAM role via AWS Console
- [ ] Check share health. Shares must become unhealthy with the message 'Principal Role is not found'
- [ ] Revoke shares and delete them
- [ ] Delete share request and consumption role
- [ ] Create another consumption role
- [ ] Create share for this new role
- [ ] Share all items,  they must be shared successfully

File: backend/dataall/modules/dataset_sharing/aws/kms_client.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/aws/kms_client.py:28]: I don't understand how the malformed principals are removed here. `AROA` principals are the IAM role canonical names, that does not mean that are "bad principals" right?

File: backend/dataall/modules/dataset_sharing/aws/s3_client.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/aws/s3_client.py:121]: Same comment as with KMS, this is not an AWS boto3 call, seems more suitable to place it in the share managers in the service layer

File: backend/dataall/modules/dataset_sharing/services/data_sharing_service.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:113]: Maybe we can place the verification of the principal role before the dataset lock. In line 79 `lock_acquired = cls.acquire_lock_with_retry(dataset.datasetUri, session, share.shareUri)` we might be waiting and locking the dataset in vain for a share that is going to fail either way

File: backend/dataall/modules/dataset_sharing/services/share_object_service.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_object_service.py:71]: This function is doing 2 things: it is verifying the principal IAM role and then it is updating the health status, this second part might be problematic. Health-statuses are only changed in the verification workflows of the item. If we start adding changes in other parts of the code, it will soon start be difficult to follow (very similar to what happened with the regular share workflows in the past).

File: backend/dataall/modules/dataset_sharing/services/share_object_service.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_object_service.py:61]: We do not need to send the delegation_role as it defaults to it

File: backend/dataall/modules/dataset_sharing/services/share_object_service.py
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/services/share_object_service.py:71]: With that in mind, I would break it down and only do the verification of the "does IAM role exist?" here. Which leads to the idea of, do we even need to make it a service function of sharing? We already have `base/aws.iam.IAM.get_role_by_name` that we can call directly, and as explained before it is just a:

```
environment = EnvironmentService.get_environment_by_uri(session, share.environmentUri)
principal_role = IAM.get_role_arn_by_name(
                account=environment.AwsAccountId,
                region=environment.region,
                role_name=share.principalIAMRoleName
                )
```

**Comment by dlpzx** [comment]
Let me first confirm my understanding on how this PR interacts with the current workflows. For data sharing, the Share.Status and ShareItem.Status should move in this statemachine:
image

- `AddItem` = No changes, it is alright if the users add items, it is just info in RDS in `pending state`
- `Submit` => Should we allow requesters to submit a request when the IAM role does not exist?
- `Approve` => Should we allow approvers to approve a request when the IAM role does not exist?
- `Reject` => No changes, it is alright if the users reject, it is just info in RDS in `rejected`
- One the share ECS task `starts` => yes, as first item it should check if the principal IAM role exists:
      -  if the task is a `approve_share` task and the IAM role does not exist => treat is as another `Failure` (more or less what is implemented at the moment)
      - 👀  if the task is a `revoke_share` => clean-up the rest of the share and ensure the revoke completes => This is the most important part of the PR. If needed we could just implement this item. 👀 
- `RevokeItems` = No changes, it is alright if the users revokes items, data.all must ensure the share revoke won't fail, same for `DeleteItems`

File: backend/dataall/modules/dataset_sharing/services/share_managers/s3_bucket_share_manager.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_managers/s3_bucket_share_manager.py:456]: I don't think we need to pass the `principal_exist`, we can just use the function `IAM.get_role_arn_by_name` if it returns `None` then the principal does not exist. Like what is implemented in line 549

File: backend/dataall/modules/dataset_sharing/services/share_managers/s3_access_point_share_manager.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_managers/s3_access_point_share_manager.py:38]: I think there are a couple of places where a share revoke could fail here. For example in `delete_dataset_bucket_policy` we call:
```
target_requester_arn = IAM.get_role_arn_by_name(
            self.target_account_id, self.target_environment.region, self.target_requester_IAMRoleName
        )
```
Which won't return anything, I would verify all calls to ` IAM.get_role_arn_by_name` in the data sharing functions

File: backend/dataall/modules/dataset_sharing/services/share_processors/s3_access_point_process_share.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_processors/s3_access_point_process_share.py:132]: Instead of passing the `principal_exist` variable I would check directly where it is used, so internally in `revoke_target_role_access_policy`

Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/services/share_processors/s3_access_point_process_share.py:132]: Checking `revoke_target_role_access_policy` I realize that even if the role does not exist, the policy will exist and we need to clean it up

File: backend/dataall/modules/dataset_sharing/services/share_processors/s3_bucket_process_share.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_processors/s3_bucket_process_share.py:92]: Minor. The exception message is a bit misleading. We can only revoke it has been approved. For users it is impossible to revoke this share

File: backend/dataall/modules/dataset_sharing/services/share_processors/s3_bucket_process_share.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_processors/s3_bucket_process_share.py:129]: Similar to the comment in the access point shares, I suggest we move the parameter calculation closer to where it is actually used instead of passing it from the top. (check if principal exists directly inside `delete_target_role_bucket_policy`)

File: backend/dataall/modules/dataset_sharing/aws/kms_client.py
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/aws/kms_client.py:28]: For these policies everywhere in dataall we use Arns to identify the principal. 
But if the principal is deleted, automatically (i think, it's IAM behaviour) those Arns are replaced by RoleIds.  So, if in data-all created S3 (and KMS) policies you encounter RoleId instead of Arn, it 99% percent means, that the role was deleted.

Of course, to be 100% sure, In case of Malformed policy exception, I need to check all principals by their IDs. Unfortunately boto3 has no instrument to list roles by RoleId (only aws cli has). So, in order to make thinfs easier, I made it that way.

File: backend/dataall/modules/dataset_sharing/aws/kms_client.py
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/aws/kms_client.py:12]: I thought about it. 
BUT
1. This method is only used here.
2. We call create_policy in many places, so, it's easy to forget to call _remove_malformed_principal before it. Here, on the other hand, it makes sense: if we have ClientError, we try to fix it.

I don't insist, but I think, it's more elegant to leave it here.

**Comment by SofiaSazonova** [comment]
@dlpzx 
> * `Submit` => Should we allow requesters to submit a request when the IAM role does not exist?
> * `Approve` => Should we allow approvers to approve a request when the IAM role does not exist?

Good point, we can terminate it right there. 

All other cases are already covered in PR.

File: backend/dataall/modules/dataset_sharing/services/share_managers/s3_access_point_share_manager.py
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/services/share_managers/s3_access_point_share_manager.py:38]: Here it doesn't matter if it is None or not. The further actions are the same

Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/services/share_managers/s3_access_point_share_manager.py:39]: Issue adressed in [30b9d65](https://github.com/data-dot-all/dataall/pull/1161/commits/30b9d65b21969c27d2cea993bfd0bf2a86fa360d)

File: backend/dataall/modules/dataset_sharing/services/share_processors/s3_access_point_process_share.py
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/services/share_processors/s3_access_point_process_share.py:132]: Well, surprisingly, if I delete role via console, all policies are also deleted with it.
I am not sure, that we will be able to delete this role and keep policies

Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/services/share_processors/s3_access_point_process_share.py:132]: Anyway, it won't hurt to check

File: backend/dataall/modules/dataset_sharing/services/share_processors/s3_bucket_process_share.py
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/services/share_processors/s3_bucket_process_share.py]: adressed in [b60f12d](https://github.com/data-dot-all/dataall/pull/1161/commits/b60f12d84c0cef198310dc6b022753d925816b7c)

File: backend/dataall/modules/dataset_sharing/services/share_processors/s3_access_point_process_share.py
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/services/share_processors/s3_access_point_process_share.py]: adressed in [b60f12d](https://github.com/data-dot-all/dataall/pull/1161/commits/b60f12d84c0cef198310dc6b022753d925816b7c)

File: backend/dataall/modules/dataset_sharing/services/share_managers/s3_bucket_share_manager.py
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/services/share_managers/s3_bucket_share_manager.py]: adressed in [b60f12d](https://github.com/data-dot-all/dataall/pull/1161/commits/b60f12d84c0cef198310dc6b022753d925816b7c)

File: backend/dataall/modules/dataset_sharing/services/share_object_service.py
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/services/share_object_service.py]: adressed in [6e5951e](https://github.com/data-dot-all/dataall/pull/1161/commits/6e5951e7a849cdb0230f27379d3523fe2ace96b8)

File: backend/dataall/modules/dataset_sharing/services/share_object_service.py
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/services/share_object_service.py]: adressed in [e6d8](https://github.com/data-dot-all/dataall/pull/1161/commits/127e6d87e4368ce0c8e761f0ac7a66ab6ae04da9)

File: backend/dataall/modules/dataset_sharing/services/share_processors/s3_bucket_process_share.py
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/services/share_processors/s3_bucket_process_share.py:93]: addressed in [d216e9a](https://github.com/data-dot-all/dataall/pull/1161/commits/d216e9a3cc8e9024b759afadae86d6a20145582b)

File: backend/dataall/modules/dataset_sharing/aws/kms_client.py
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/aws/kms_client.py:28]: Initially we used to write principals with roleIds, but we moved to arns as they are easier to read and understand by end-users. That's why I thought you were changing this. Because we used to work with role ids a lot, there are/were methods to get the role ids directly in the code. You can read the role id from a role by calling the IAM boto3 client:

```
    def get_role_id(accountid, name):
        session = SessionHelper.remote_session(accountid=accountid)
        client = session.client('iam')
        try:
            response = client.get_role(RoleName=name)
            return response['Role']['RoleId']
        except ClientError:
            return None
```
I think we should not clean-up all role ids because the majority of customers use "imported" Buckets and KMS keys and we can wipe out permissions set outside of data.all.

File: backend/dataall/modules/dataset_sharing/services/data_sharing_service.py
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:113]: The lock is for the dataset resources, we want to prevent that the Bucket and KMS policies being updated during a share task are being updated simultaneously by another share task. A share item is unique for each share, there is no way another share is simultaneously updating it, plus RDS supports ACID transactions no?

File: backend/dataall/modules/dataset_sharing/aws/kms_client.py
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/aws/kms_client.py:12]: Let me have a look at how the code looks after changes because it might be good now. I would be strict with the definition of layers, because we can end up mixing responsibilities and having a code that are just exceptions to the rule

Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/aws/kms_client.py:12]: So, we are calling `put_key_policy` and that's when in case of MalformedPolicy we remove the malformed principals. We are doing all the policy manipulation in the s3_X_share_managers except for this piece that is a bit hidden in the aws layer. I propose to go for:
- kms client = put_key_policy that just calls boto3 KMS put_key_policy and rasies errors

There are 4 places (+tests) where we call that function, we could insert there the logic of put_policy -> raise exception -> remove malformed -> try again to put_policy. e.g. `dataall/modules/dataset_sharing/services/share_managers/s3_access_point_share_manager.py:723`

```
                try:
                    kms_client.put_key_policy(kms_key_id, json.dumps(existing_policy))
                except 'MalformedPolicyDocumentException': #catch error better
                    clean_policy = self._remove_malformed_principals(existing_policy)
                    kms_client.put_key_policy(kms_key_id, json.dumps(clean_policy))
```

File: backend/dataall/modules/dataset_sharing/services/share_managers/s3_access_point_share_manager.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_managers/s3_access_point_share_manager.py:538]: typo: KMS key policy

File: backend/dataall/modules/dataset_sharing/services/share_processors/s3_bucket_process_share.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_processors/s3_bucket_process_share.py:92]: Why do we add this check in the buckets `process_approved_shares` but we do not do it for access points or tables? I would make it fail before triggering the ECS task, like the next steps of adding a check in the approve of the share

File: backend/dataall/modules/dataset_sharing/services/data_sharing_service.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:598]: We cannot update all share items `HealthStatus`  to `Unhealthy`, we can only update those whose `Status` is in `Share_Succeed`. Those items in other sharing states (e.g. `PendingApproval` do not have a share to be healthy or unhealthy)

File: backend/dataall/modules/dataset_sharing/services/data_sharing_service.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:479]: IMPORTANT: In approve shares and revoke shares we are procssing share requests, we are not verifying the healthstatus (check the big comment with the picture). In approve_share, if the role is not present, the `Status` of the item should go to failure. We do not manage the unhealthyness of a share, because it has not happened yet. We could argue that it is our chance to update the customer and tell them that OTHER share items already in `Share_Succeed` are unhealthy

**Review by dlpzx — CHANGES_REQUESTED** [review]
Some comments and one important note on what is the shareStatus, healthStatus lifecycle. The big comment with the picture contains all the details

File: backend/dataall/modules/dataset_sharing/aws/kms_client.py
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/aws/kms_client.py:28]: Another option: store consumption role IDs in DB. What do you think?

File: backend/dataall/modules/dataset_sharing/services/data_sharing_service.py
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:479]: Update Health message and status is our only way to show the user in UI, WHY did share fail. Otherwise it will be unclear.

File: backend/dataall/modules/dataset_sharing/aws/kms_client.py
Comment (SofiaSazonova) [reply backend/dataall/modules/dataset_sharing/aws/kms_client.py:28]: On the other hand, there can be several deleted roles at once, so it's better to wipe them all by checking principal ID.

File: backend/dataall/modules/dataset_sharing/aws/s3_client.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/aws/s3_client.py:131]: This comment is outdated

File: backend/dataall/modules/dataset_sharing/db/share_object_repositories.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/db/share_object_repositories.py:824]: small typo: shareable

File: backend/dataall/modules/dataset_sharing/services/data_sharing_service.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:157]: Very minor: principal_healthy is a weird variable name to represent the success of the shares

File: backend/dataall/modules/dataset_sharing/services/share_object_service.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/share_object_service.py:68]: We should only sent to Unhealthy those items that are currently shared. If an item is in PendingState status it should not be in any state yet

File: backend/dataall/modules/dataset_sharing/services/share_processors/s3_bucket_process_share.py
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/services/share_processors/s3_bucket_process_share.py:92]: ^^

**Review by dlpzx — COMMENTED** [review]
Some additional comments

File: backend/dataall/modules/dataset_sharing/services/data_sharing_service.py
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:475]: After discussions offline, we have decided to implement better UX for share failed errors in https://github.com/data-dot-all/dataall/issues/1168

File: backend/dataall/modules/dataset_sharing/services/data_sharing_service.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:82]: typo: successful

File: backend/dataall/modules/dataset_sharing/services/data_sharing_service.py
Comment (dlpzx) [inline backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:585]: Move check in line 587 before the dataset_lock logic --> to line 553

**Comment by dlpzx** [comment]
I think we need to add some test cases

- [ ] Create consumption role for the env and share request for this role ---> all items are shared successfully
- [ ] Now, delete the IAM role and click verify items and verify some of the items. ---> The verified items only must become unhealthy with the message 'Principal Role is not found'
- [ ] Revoke all items from the share ---> the unhealthy as well as the healthy items must be revoked successfully
       - [ ] Removed from KMS policy
       - [ ] Removed from S3 policy
       - [ ] Managed IAM policies deleted
- [ ] Re-add the items and submit the share request --> When share is submitted it should throw an error
- [ ] Re-create the consumption role, re-add the items, submit the request, delete the consumption role and the approvers click on approve --> When share is approved, it should throw an error

**Comment by SofiaSazonova** [comment]
all these tests were made successfully)

**Review by dlpzx — APPROVED** [review]
Looks great! Thanks!

**Pull request merged**
